docs/rfc/block-registration.md

```diff
@@ -0,0 +1,523 @@
+
+**Example:**
+
+
+```json
+{
+	"name": "my-plugin/notice",
+	"title": "Notice",
+	"category": "common",
+	"parent": [ "core/group" ],
+	"icon": "star",
+	"description": "Shows warning, error or success notices  ...",
+	"keywords": [ "alert", "message" ],
+	"textDomain": "my-plugin",
+	"attributes": {
+		"message": {
+			"type": "string",
+			"source": "html",
+			"selector": ".meessage"
+		}
+	},
+	"styles": [ 
+		{ "name": "default", "label": "Default", "isDefault": true }, 
+		{ "name": "other", "label": "Other" }
+	],
+	"editorScript": "file:build/editor.assets.json",
+	"script": {
+		"handle": "my-plugin-notice",
+		"file": "index.js",
+		"dependencies": [ "wp-element", "wp-components" ],
+		"version": "3.0.0"
+	},
+	"editorStyle": {
+		"handle": "my-plugin-notice-editor",
+		"file": "editor.css",
+		"dependencies": [ "wp-edit-blocks" ]
+	},
+	"style": "my-plugin-notice",
+	"renderCallback": "file:my-render-callback.php"
+}
+```
+
+## Block API
+
+This section describes all the properties that can be added to the `block.json` file to define the behavior and metadata of block types.
+
+### Name
+
+* Type: `string`
+* Required
+* Localized: No
+* Property: `name`
+
+```json
+{ "name": "core/heading" }
+```
+
+The name for a block is a unique string that identifies a block. Names have to be structured as `namespace/block-name`, where namespace is the name of your plugin or theme.
+
+**Note:** A block name can only contain lowercase alphanumeric characters, dashes, and at most one forward slash to designate the plugin-unique namespace prefix. It must begin with a letter.
+
+**Note:** This name is used on the comment delimiters as `<!-- wp:my-plugin/book -->`. Block types in the `core` namespace do not include a namespace when serialized.
+
+### Title
+
+* Type: `string`
+* Required
+* Localized: Yes
+* Property: `title`
+
+```json
+{ "title": "Heading" }
+```
+
+This is the display title for your block, which can be translated with our translation functions. The block inserter will show this name.
+
+### Category
+
+* Type: `string`
+* Required
+* Localized: No
+* Property: `category`
+
+```json
+{ "category": "common" }
+```
+
+Blocks are grouped into categories to help users browse and discover them.
+
+The core provided categories are:
+
+* common
+* formatting
+* layout
+* widgets
+* embed
+
+Plugins and Themes can also register [custom block categories](/docs/designers-developers/developers/filters/block-filters.md#managing-block-categories).
+
+An implementation should expect and tolerate unknown categories, providing some reasonable fallback behavior (e.g. a "common" category).
+
+### Parent
+
+* Type: `string[]`
+* Optional
+* Localized: No
+* Property: `parent`
+
+```json
+{ "parent": [ "my-block/product" ] }
+```
+
+Setting `parent` lets a block require that it is only available when nested within the specified blocks. For example, you might want to allow an 'Add to Cart' block to only be available within a 'Product' block.
```

Comment:
Are wildcards permitted here?

Reply:
@noisysocks or @jorgefilipecosta, do you remember whether wildcards will work as well? I know it was discussed during the implementation phase.

Reply:
This is the effective implementation:


https://github.com/WordPress/gutenberg/blob/0e46e9ef92d27aa4af849c482c81216b9c579410/packages/block-editor/src/store/selectors.js#L1092-L1100

To answer the question: No, wildcards aren't (currently) supported.

Separately:

- A boolean type _is_ supported, which isn't reflected yet in this proposal.
- If and when #14908, there will be precedent for wildcards in block transforms ([source](https://github.com/WordPress/gutenberg/pull/14908/files#diff-a5c77b070da74373cd1d937dee3a77eaR242)), and it certainly seems like an equivalent behavior _could_ be implemented for `parent`.

Reply:
That's right. From memory, we decided to keep it very barebones with a view to iterating on it.

Reply:
Should we include `boolean` here as well? It looks like `true` is something which you would never want to use as it means - allow all. `false` can be represented as `[]`.

Reply:
> * If and when #14908, there will be precedent for wildcards in block transforms ([source](https://github.com/WordPress/gutenberg/pull/14908/files#diff-a5c77b070da74373cd1d937dee3a77eaR242)), and it certainly seems like an equivalent behavior _could_ be implemented for `parent`.

We should create a follow-on issue for this. That said, it's not directly mentioned as being possible in this RFC. Which begs the question: Should it be, if we're saying it ought to be possible?

Reply:
Copied to #16209 which contains the list of future enhancements to this RFC.